Set current update lane priority for user blocking events

rickhanlonii opened the pull request:
## Summary

We missed these callsites in https://github.com/facebook/react/pull/19121 and one is the most common mismatch right now.

acdlite commented:
The feature flag should be static since it only exists to prevent it from leaking into open source. Can be static `true` in FB dev builds since we use the warning filter to gate.

acdlite commented on packages/react-dom/src/events/ReactDOMEventListener.js:
```diff
@@ -148,16 +149,19 @@ function dispatchUserBlockingUpdate(
-    ),
-  );
+  // TODO: Double wrapping is necessary while we decouple Scheduler priority.
+  unstable_runWithPriority(UserBlockingPriority, () => {
```
Since this is part of the renderer, we should be able to mutate the value directly without a wrapper function.

Not gonna block the PR over this, but does need to be fixed before we can land the experiment..

acdlite replied:
Wait, also, isn't UserBlockingPriority the wrong type? Should be a lane priority.

rickhanlonii replied:
> isn't UserBlockingPriority the wrong type? Should be a lane priority.

It must have been buried, but I mentioned this in the comment [here](https://github.com/facebook/react/pull/19121#discussion_r449238705) - ReactDOM.runWithPriority accepts a `ReactPriorityLevel` instead of a `LanePriority` so that we don't have to leak the lane implementation while we do the decoupling. Internally in React, everything operates on the Lane priority.

Were you expecting that we would export all of the LanePriorities, and the `schedulerPriorityToLanePriority` conversion function from ReactFiberReconciler so we can pass LanePriorities to ReactDOM.runWithPriority?

> Since this is part of the renderer, we should be able to mutate the value directly without a wrapper function.

Do you mean that we wouldn't need the try/catch? That would require that we export the current update lane getter and setter from ReactFiberReconciler as well as all the LanePriorities like mentioned above. Right now we only export `runWithPriority (ReactPriorityLevel)` and `getCurrentUpdatePriority (ReactPriorityLevel)`.

acdlite replied:
I think it's fine to leak the lane priorities to the renderer because it's part of the same build artifact. It doesn't cross any package boundary.

Leaking to via an API `ReactDOM.runWithPriority` is different because that's called by external code.

> ReactDOM.runWithPriority accepts a ReactPriorityLevel instead of a LanePriority so that we don't have to leak the lane implementation while we do the decoupling. Inside the reconciler, everything operates on the Lane priority.

I missed that comment originally but I don't buy that rationale. `ReactDOM.runWithPriority` is an unstable API that only exists internally if/when we need it to temporarily migrate external callers. We're going to delete it ASAP, so "leaking" update lane priorities is not a huge issue. But more importantly, the caller in this PR isn't external, so that's not relevant in this case.

rickhanlonii commented:
@acdlite I replaced runWithPriority with direct updates, and updated the exposed methods to use LanePriority. I can follow up with the feature flags in the next diff so this is easier to review if you want to re-review this now.

rickhanlonii commented:
The next branch is here: https://github.com/rickhanlonii/react/compare/rh-add-rwp-to-events...rickhanlonii:rh-ff-update-lane?expand=1

acdlite commented on packages/react-dom/src/events/ReactDOMEventListener.js:
```diff
@@ -148,6 +153,9 @@ function dispatchUserBlockingUpdate(
 ) {
+  // TODO: Double wrapping is necessary while we decouple Scheduler priority.
+  const previousPriority = getCurrentUpdateLanePriority();
+  setCurrentUpdateLanePriority(InputContinuousLanePriority);
```
Needs a try-catch, since runWithPriority could throw

rickhanlonii replied:
I see, I misunderstood "we should be able to mutate the value directly without a wrapper function" as not needing a try/catch.

acdlite commented on packages/react-dom/src/events/DeprecatedDOMEventResponderSystem.js:
```diff
@@ -101,9 +108,12 @@ const eventResponderContext: ReactDOMResponderContext = {
         runWithPriority(UserBlockingPriority, () =>
           executeUserEventHandler(eventListener, eventValue),
         );
+        setCurrentUpdateLanePriority(previousPriority);
```
Needs a try-catch

acdlite commented on packages/react-dom/index.experimental.js:
```diff
@@ -23,8 +23,9 @@ export {
-  // DO NOT USE: Temporarily exposing this to migrate off of Scheduler.runWithPriority.
+  // DO NOT USE: Temporarily exposing these to migrate off of Scheduler.runWithPriority.
   unstable_runWithPriority,
+  unstable_IdleLanePriority,
```
Just hardcode the number 2 instead of importing.

acdlite commented on packages/react-noop-renderer/src/ReactNoop.js:
```diff
@@ -47,9 +47,9 @@ export const {
-  // This should only be used by React internals.
+  // TODO: Remove these after callers migrate to alternatives.
   unstable_runWithPriority,
+  unstable_InputContinuousLanePriority,
```
Same comment as before. These are only used internally. Hardcode the number wherever it’s used, with a comment and a name that’s easily searchable.

gaearon replied:
You can remove this now?

gaearon commented on packages/react-noop-renderer/src/createReactNoop.js:
```diff
@@ -949,6 +949,8 @@ function createReactNoop(reconciler: Function, useMutation: boolean) {
     },
 
     unstable_runWithPriority: NoopRenderer.runWithPriority,
+    unstable_InputContinuousLanePriority:
```
Same.

gaearon commented on packages/react-reconciler/src/ReactFiberLane.js:
```diff
@@ -66,7 +66,7 @@ const RetryLanePriority: LanePriority = 5;
 
 const IdleHydrationLanePriority: LanePriority = 3;
-const IdleLanePriority: LanePriority = 2;
+export const IdleLanePriority: LanePriority = 2;
```
This change isn't needed now?
